fields/types/gcsimage/GcsImageType.js

```diff
@@ -282,79 +282,40 @@ gcsimage.prototype.uploadFile = function (item, file, update, callback) {
-				// resizing image and upload resized images
-			if (typeof _this.options.resize === 'function') {
-				var resizeFunc = _this.options.resize;
-				if (Array.isArray(_this.options.resizeOpts) && _this.options.resizeOpts.length > 0) {
-					var promises = [];
-					var targets = {};
-					var resizeOpts = _this.options.resizeOpts;
-					for (var i = 0; i < resizeOpts.length; i++) {
-						var resizeOpt = resizeOpts[i] || {};
-						targets[resizeOpt.target] = {
-							url: gcsHelper.getPublicUrl(_this.options.bucket, gcsDir + filenameWithoutExt + '-' + resizeOpt.target + '.' + ext),
-							width: resizeOpt.width,
-							height: resizeOpt.height,
-						};
-
-                        _.set(resizeOpt, 'options.filetype', filetype)
-
-						promises.push(gcsHelper.uploadFileToBucket(bucket, resizeFunc(file.path, resizeOpt.width, resizeOpt.height, resizeOpt.options), {
-							destination: gcsDir + filenameWithoutExt + '-' + resizeOpt.target + '.' + ext,
-							filetype: filetype,
-							isPublicRead: isPublicRead,
-							cacheControl: 'public, max-age=' + ONE_YEAR,
-						}));
-					}
-					return Promise.all(promises).then(function (values) {
-						return targets;
-					});
-				} else {
-						// skip resizing
-					return;
-				}
-			} else {
-					// skip resizing
-				return;
-			}
-		}).then(function (targets) {
-			var dimensions = sizeOf.sync(fs.readFileSync(file.path));
-
-			if (targets) {
-				try {
-					_.forEach(targets, function (target) {
-						if (typeof target.width === 'number' && !target.height) {
-							if (target.width < dimensions.width) {
-								target.height = Math.round((target.width / dimensions.width) * dimensions.height);
-							} else {
-								target.width = dimensions.width;
-								target.height = dimensions.height;
-							}
-						} else if (typeof target.height === 'number' && !target.width) {
-							if (target.height < dimensions.height) {
-								target.width = Math.round((target.height / dimensions.height) * dimensions.width);
-							} else {
-								target.width = dimensions.width;
-								target.height = dimensions.height;
-							}
-						}
-					});
-				} catch (e) {
-					console.warn('Calculating width and height of resized image occurs error ', e);
-				}
-			}
+		}).then(function () {
+			const dimensions = sizeOf.sync(fs.readFileSync(file.path));
+
+      const targets = {};
+
+      [['desktop', 2000], ['tablet', 1200], ['mobile', 800], ['w400', 400], ['tiny', 150]].forEach(resizeOpt => {
+        let [resizeName, resizeWidth] = resizeOpt;
+        let resizeHeight = 0;
+
+        if (resizeWidth < dimensions.width) {
+          resizeHeight = Math.round((resizeWidth / dimensions.width) * dimensions.height);
+        } else {
+          resizeWidth = dimensions.width;
+          resizeHeight = dimensions.height;
+        }
+
+        targets[resizeName] = {
+          url: gcsHelper.getPublicUrl(_this.options.bucket, gcsDir + filenameWithoutExt + '-' + resizeName[0] + '.' + ext),
```

Comment:
Is it possible to use different buckets for the raw photo and the resized ones? As we don't reveal our original photo in news_v2 endpoints, do we still the need put them within the same bucket?  In GCF, the resize event is triggered by the creation of the object within the bucket after uploading directly. If these photos reside in the same bucket, the resized photos will incur redundant events. Extra validation of the file types (e.g, photos without -desktop.JPG -w400.JPG are the target raw photo) that should be taken within the cloud function which might not be straightforward.

Reply:
Sure, we can create a new bucket for the original images.

By the way, in this PR, we don't resize the images. 
However, we still store the resize information into database.
Could we add those resize information in the Google Cloud Function?
Therefore, we don't need to specify two different buckets here:
one for uploading original images;
and the other is for storing the resize information.

Reply:
In current design, we generate the metadata information before the photos are resized. I don't think we need to change the mechanism here. If we move this change to GCF, we need to grant extra permission for it to access the mongo cluster. That is the least strategy to I'd take. What is the issue to use different buckets here?

Reply:
In current mechanism, `gcsimage` is not general anymore since we have hard codes to store the metadata.
If we want to let `gcsimage` to handle resize metadata, I would like to change current logics.

I will update the PR later after the new implementation way ready.

Reply:
After reviewing the architecture, I found that this approach would force to define the resizing options on both service [plate](https://github.com/twreporter/plate/pull/166) and [cloud function](https://github.com/twreporter/cloud-functions/blob/master/task/publish.go#L27-L33) which will complicate the system. I came up with two approaches yesterday.

1. Define the resizing option on cloud function (previous apporach)
To define the resizing options on cloud functions, we can remove the metadata injection on keystone. Delegate the cloud function to update the image records. However, to enable the cloud function to access mongodb clusters, we need to add [extra connector](https://cloud.google.com/vpc/docs/configure-serverless-vpc-access) to allow the internal access.

Pros:
* No extra dependency on keystone.

Cons:
* Require another f1-micro(connector) VM & VPC network cost
* Require extra permission to allow cloud function to access mongodb cluster
* Redundant create(resize image) event filter is required to allow single bucket usage or
  need extra migration efforts to support raw photos & resized photos bucket.


2. Define the resizing option on keystone
To define the resizing options on keystone, we can publish the resizing events directly to the pubsub according to the [format](https://github.com/twreporter/cloud-functions/blob/master/task/publish.go#L69-L77).

Pros
* Reduce the cloud function service number.
* No extra bill cost against 1.
* No

Cons
* Introduce new dependency(pubsub client) on keystone

Thanks for the consideration and let me know your thoughts.

Reply:
About Approach 2, could you help to explain more in details?
For example, wow do it update the mongodb records?

The latest commits in this PR and [plate PR](https://github.com/twreporter/plate/pull/166) are similar to Approach 2.
We don't define resize details in `Keystone` library, we define them in `plate/models/Image.js`.
We use `hooks.postUpload` to update the resize metadata into database records.

Reply:
> About Approach 2, could you help to explain more in details?
> For example, how do it update the mongodb records?
The image metadata are still processing through `hooks.postUpload` but with additional operation to publish the corresponding event.

> The latest commits in this PR and [plate PR](https://github.com/twreporter/plate/pull/166) are similar to Approach 2.
> We don't define resize details in `Keystone` library, we define them in `plate/models/Image.js`.
> We use `hooks.postUpload` to update the resize metadata into database records.
I just continued the discussion with the same thread. If you think the plate PR is more suitable for the details, we can move on to there.

Reply:
> I just continued the discussion with the same thread. If you think the plate PR is more suitable for the details, we can move on to there.

Yes, we could discuss Approach 2 in the pate PR.